rocks.ml

```diff
@@ -215,13 +224,232 @@ module rec Iterator : Rocks_intf.ITERATOR with type db := RocksDb.t = struct
+      "rocksdb_transaction_put"
+      (t @->
+       ocaml_string @-> Views.int_to_size_t @->
+       ocaml_string @-> Views.int_to_size_t @->
+       returning_error void)
+
+  let put ?(key_pos=0) ?key_len ?(value_pos=0) ?value_len ?opts t key value =
+    let open Bigarray.Array1 in
+    let key_len = match key_len with None -> dim key - key_pos | Some len -> len in
+    let value_len = match value_len with None -> dim value - value_pos | Some len -> len in
+    with_err_pointer begin
+        put_raw t
+          (bigarray_start array1 key +@ key_pos) key_len
+          (bigarray_start array1 value +@ value_pos) value_len
+      end
+
+  let put_string ?(key_pos=0) ?key_len ?(value_pos=0) ?value_len ?opts t key value =
+    let key_len = match key_len with None -> String.length key - key_pos | Some len -> len in
+    let value_len = match value_len with None -> String.length value - value_pos | Some len -> len in
+    with_err_pointer begin
+        put_raw_string t
+          (ocaml_string_start key +@ key_pos) key_len
+          (ocaml_string_start value +@ value_pos) value_len
+      end
+
+  let delete_raw =
+    foreign
+      "rocksdb_transaction_delete"
+      (t @->
+       ptr char @-> Views.int_to_size_t @->
+       returning_error void)
+
+  let delete_raw_string =
+    foreign
+      "rocksdb_transaction_delete"
+      (t @->
+       ocaml_string @-> Views.int_to_size_t @->
+       returning_error void)
+
+  let delete ?(pos=0) ?len ?opts t key =
+    let open Bigarray.Array1 in
+    let len = match len with None -> dim key - pos | Some len -> len in
+    with_err_pointer (delete_raw t (bigarray_start array1 key +@ pos) len)
+
+  let delete_string ?(pos=0) ?len ?opts t key =
+    let len = match len with None -> String.length key - pos | Some len -> len in
+    with_err_pointer (delete_raw_string t (ocaml_string_start key +@ pos) len)
+  let write_raw =
+    foreign
+      "rocksdb_write"
+      (t @-> WriteOptions.t @-> WriteBatch.t @->
+       returning_error void)
+
+  let get_raw =
+    foreign
+      "rocksdb_transaction_get"
+      (t @-> ReadOptions.t @->
+       ptr char @-> Views.int_to_size_t @-> ptr Views.int_to_size_t @->
+       returning_error (ptr char))
+
+  let get_raw_string =
+    foreign
+      "rocksdb_transaction_get"
+      (t @-> ReadOptions.t @->
+       ocaml_string @-> Views.int_to_size_t @-> ptr Views.int_to_size_t @->
+       returning_error (ptr char))
+
+  let get ?(pos=0) ?len ?opts t key =
+    let open Bigarray.Array1 in
+    let len = match len with None -> dim key - pos | Some len -> len in
+    let inner opts =
+      let res_size = allocate Views.int_to_size_t 0 in
+      let res = with_err_pointer
+          (get_raw t opts (bigarray_start array1 key +@ pos) len res_size)
+      in
+      if (to_voidp res) = null
+      then None
+      else begin
+        let res' = bigarray_of_ptr array1 (!@res_size) Bigarray.char res in
+        Gc.finalise (fun res -> free (to_voidp res)) res;
```

Comment:
Does this `free` actually need to be in a finalizer?
(I know, in other places it's done like that too ... but I'm wondering if I made a mistake there)

Reply:
Honestly, I don't know. Since C API is a bit different I was forced to copy wrappers from your code.

Reply:
hmm, I guess for the `bigarray_of_ptr` it is required to wait with `free`ing until `res'` is no longer referenced, see http://ocamllabs.io/ocaml-ctypes/Ctypes.html#VALbigarray_of_ptr
I'm not sure `res'` will also keep `res` alive ... need to dig into ctypes a bit more again.

For the `string_from_ptr` variant I think we can free the original c-string immediately.